utils/sb3/custom_ppo.py

```diff
@@ -0,0 +1,154 @@
+"""Module containing regularized PPO algorithm."""
+
+import logging
+
+import numpy as np
+import torch
+from gymnasium import spaces
+from stable_baselines3 import PPO
+from stable_baselines3.common.callbacks import BaseCallback
+from stable_baselines3.common.utils import obs_as_tensor
+from stable_baselines3.common.vec_env import VecEnv
+
+# Import masked buffer class
+from utils.sb3.masked_buffer import MaskedRolloutBuffer
+
+logging.getLogger(__name__)
+
+
+class MultiAgentPPO(PPO):
+    """Adapted Proximal Policy Optimization algorithm (PPO) that is compatible with multi-agent environments."""
+
+    def _setup_model(self) -> None:
+        super()._setup_model()
+
+        # Change buffer to our own masked version
+        buffer_cls = MaskedRolloutBuffer
+
+        self.rollout_buffer = buffer_cls(
+            self.n_steps,
+            self.observation_space,  # type: ignore[arg-type]
+            self.action_space,
+            device=self.device,
+            gamma=self.gamma,
+            gae_lambda=self.gae_lambda,
+            n_envs=self.n_envs,
+        )
+
+    def collect_rollouts(
+        self,
+        env: VecEnv,
+        callback: BaseCallback,
+        rollout_buffer: MaskedRolloutBuffer,
+        n_rollout_steps: int,
+    ) -> bool:
+        """Adapted collect_rollouts function."""
+
+        assert self._last_obs is not None, "No previous observation was provided"
+        # Switch to eval mode (this affects batch norm / dropout)
+        self.policy.set_training_mode(False)
+
+        n_steps = 0
+        rollout_buffer.reset()
+        # Sample new weights for the state dependent exploration
+        if self.use_sde:
+            self.policy.reset_noise(env.num_envs)
+
+        callback.on_rollout_start()
+
+        while n_steps < n_rollout_steps:
+            if self.use_sde and self.sde_sample_freq > 0 and n_steps % self.sde_sample_freq == 0:
+                # Sample a new noise matrix
+                self.policy.reset_noise(env.num_envs)
+
+            with torch.no_grad():
+                # Convert to pytorch tensor or to TensorDict
+                obs_tensor = obs_as_tensor(self._last_obs, self.device)
+
+                # EDIT_1: Mask out invalid observations (NaN dimensions and/or dead agents)
+                # Create dummy actions, values and log_probs (NaN)
+                actions = torch.full(fill_value=np.nan, size=(self.n_envs,))
+                log_probs = torch.full(fill_value=np.nan, size=(self.n_envs,))
+                values = torch.full(fill_value=np.nan, size=(self.n_envs,)).unsqueeze(dim=1)
+
+                for idx, agent_id in enumerate(env.agent_ids):
+                    if agent_id not in env.dead_agent_ids:
+                        # Sample actions from policy if agent is alive
+                        obs_tensor_agent_id = obs_tensor[idx, :].unsqueeze(dim=0)
+                        actions[idx], values[idx], log_probs[idx] = self.policy(obs_tensor_agent_id)
```

Comment:
I think this can be batched?

Reply:
Comment for myself to vectorize this; 1) Get all active agent indices; 2) use as a mask within self.policy() and store